### PR TITLE
Fix traceback on Abuse Report.

### PR DESCRIPTION
**Pull request by maranda**
The callback function at https://github.com/matrix-org/mjolnir/blob/main/src/Mjolnir.ts#L1026 receives the event packed object as single argument and not all the arguments 'emselves. Also to avoid any possible scope issues it's best to pass the Mjolnir object itself to the CB.

### CODE REVIEW

**Comment by jesopo** [comment]
did you figure out how `this` was being the wrong thing?

**Comment by maranda** [comment]
@jesopo I'm not entirely sure, I hypotize from the traceback `this` was scoped as `ReportManager` considering the handleReport() was called as callback for the event (https://github.com/matrix-org/mjolnir/blob/main/src/Mjolnir.ts#L237). And after that I also figured that the CB was being passed the unpacked event object which was another stopper.

**Comment by Yoric** [comment]
You are right that passing `this.handleReport` as callback is not a good idea in JS.

File: src/Mjolnir.ts [hunk 1]
```diff
@@ -1023,9 +1023,9 @@ export class Mjolnir {
         return await this.eventRedactionQueue.process(this, roomId);
     }
 
-    private async handleReport(roomId: string, reporterId: string, event: any, reason?: string) {
-        for (const protection of this.enabledProtections) {
-            await protection.handleReport(this, roomId, reporterId, event, reason);
+    private async handleReport(e: any) {
```
Comment (Yoric) [inline src/Mjolnir.ts:1026]: Please avoid replacing something that is (somewhat) reasonably typed by a `any`, this makes code harder to maintain.

**Review by Yoric — COMMENTED** [review]
To fix the `this`, the easiest way is to replace

```ts
reportManager.on("report.new", this.handleReport);
```

with

```ts
reportManager.on("report.new", this.handleReport.bind(this));
```

File: src/Mjolnir.ts (hunk 1 above)
Comment (maranda) [reply src/Mjolnir.ts:1026]: Hmm okay, but I'm not sure if the event object has an explicit fixed type cast there for the dictionary.

File: src/Mjolnir.ts [hunk 2]
```diff
@@ -1023,9 +1023,9 @@ export class Mjolnir {
         return await this.eventRedactionQueue.process(this, roomId);
     }
 
-    private async handleReport(roomId: string, reporterId: string, event: any, reason?: string) {
-        for (const protection of this.enabledProtections) {
-            await protection.handleReport(this, roomId, reporterId, event, reason);
+    private async handleReport(e: { mjolnir: Mjolnir, roomId: string, reporterId: string, event: any, reason?: string }) {
```
Comment (Yoric) [inline src/Mjolnir.ts:1026]: I don't think we need `mjolnir: Mjolnir` anymore, do we?

Now, `this` should work nicely.

Comment (maranda) [reply src/Mjolnir.ts:1026]: Apparently using `this` and not passing the Mjolnir object through the event still has the traceback happening. Just tested.

**Comment by Yoric** [comment]
@maranda What's the status on this?

**Comment by maranda** [comment]
@Yoric Finished and applied all suggested review changes.

**Comment by Yoric** [comment]
Adding to my review list, then!

**Review by Yoric — APPROVED** [review]
Looks good to me, thanks!

**Pull request merged**